### PR TITLE
feat(provider): store a provider API key from the command line

### PR DESCRIPTION
**Pull request by CybotTM**
Closes #581

## What this adds

`nrllm:provider:set-key <provider>` stores a provider API key from a script, doing what the setup wizard does from the backend:

```bash
printf '%s' "$OPENAI_API_KEY" | vendor/bin/typo3 nrllm:provider:set-key openai
```

Three decisions worth stating, all recorded in [ADR-124](Documentation/Adr/Adr124ProviderKeyFromTheCommandLine.rst):

- **STDIN only.** An argument would be visible in the process list and recorded in the shell history, and no option flag takes that back afterwards. A terminal is refused rather than read — prompting would hang a provisioning script in a way that looks like a freeze, so it fails visibly instead.
- **Re-running replaces, it does not re-issue.** When the provider already references a stored credential, the secret is rotated under the existing identifier. That matters because `providers.openai.apiKeyIdentifier` in the extension configuration — which the specialized speech and image services read via `ExtensionConfiguration::get('nr_llm')` — refers to that identifier and would otherwise be left pointing at a stale secret. A provider whose identifier the vault no longer knows is re-stored under that same identifier rather than given a new one.
- **`schedulable: false`.** The scheduler cannot supply STDIN.

## The second commit: a bug found on the way

The wizard passes provenance to nr-vault like this:

```php
$this->vaultService->store($vaultIdentifier, $providerApiKey, [
    'table' => 'tx_nrllm_provider',
    'field' => 'api_key',
    'source' => 'setup_wizard',
]);
```

nr-vault's `VaultService::collectOptionalFields()` reads provenance from the `metadata` option and ignores every key it does not recognise. All three were therefore discarded on every wizard run — the provenance the code appears to record has never been stored. `40df12f1` nests them.

This is in scope rather than a follow-up because the issue names "wizard and script produce different shapes" as one of the two problems to solve; shipping the new command with correct metadata while leaving the wizard silently dropping its own would have kept exactly that split. Secrets stored before this change stay unlabelled — nothing reads the metadata for behaviour, so there is no migration.

## Tests

Nine unit tests in `Tests/Unit/Command/SetProviderApiKeyCommandTest.php`, one per acceptance criterion plus the edge cases: first write, replacement keeping the identifier, a dangling identifier being re-stored, unknown provider, empty STDIN, trailing-newline handling, a failed vault write leaving the provider untouched, and two asserting the secret never reaches the output.

The input stream is built in the test rather than through `CommandTester::setInputs()`: the tester creates its input inside `execute()`, so a stream set beforehand is discarded — and an unset stream would fall through to the real STDIN of the test run.

`Tests/Unit/Fixture/InMemoryVaultService.php` is a new fixture that keeps `store` and `rotate` apart so a test can tell a first write from a replacement.

## Verification

All six pre-push suites, run on **PHP 8.2** — the version `ci.yml` pins for Rector and the floor of the test matrix:

```
unit               SUCCESS  (5758 tests, 16023 assertions)
functional sqlite  SUCCESS
phpstan            SUCCESS  [OK] No errors        (level 10)
cgl                SUCCESS
rector -n          SUCCESS  [OK] Rector is done!
fuzzy              SUCCESS
architecture       SUCCESS  (PHPat, extra — the command injects a repository)
```

unit and phpstan additionally pass on PHP 8.4.

One note for whoever runs these next: `rector -n` on PHP 8.4 reports 55 files, on 8.2 it reports none. The PHPUnit-related rules resolve against the PHPUnit major that the PHP version pulls in, so a local run on the wrong version invents a 53-file backlog that CI does not see. `ci.yml` sets `rector-php-version: '8.2'` for that reason.

### CODE REVIEW

File: Configuration/Services.yaml [hunk 1]
```diff
@@ -473,6 +473,14 @@ services:
         description: 'Run a golden question set against a retriever and report hit rates and regression.'
         schedulable: false
 
+  # Reads the secret from STDIN, so it cannot run from the scheduler (ADR-124).
+  Netresearch\NrLlm\Command\SetProviderApiKeyCommand:
+    tags:
+      - name: 'console.command'
+        command: 'nrllm:provider:set-key'
+        description: 'Store a provider API key read from STDIN and link it to the provider record.'
+        schedulable: false
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply Configuration/Services.yaml:482]: Keeping both is deliberate, and required here. TYPO3 reads `schedulable` from the DI tag only — `CommandRegistry::getSchedulableCommands()` filters on `$configuration['schedulable'] ?? true`, where $configuration is the tag attributes. `#[AsCommand]` has no way to express it, so dropping the tag would put a command that reads STDIN into the scheduler's list, where it can only hang.

Dropping the attribute instead is possible but would diverge from the repo: `EvalRunCommand` and `RetrievalEvalRunCommand` both carry `#[AsCommand]` plus an explicit tag, for the same reason (`schedulable: false`). The attribute serves Symfony, the tag serves TYPO3's registry.

The drift risk is real but narrow: the tag's `command` key is what TYPO3 dispatches on, so a divergent name in the attribute cannot silently route somewhere else — it would fail visibly.

File: Classes/Command/SetProviderApiKeyCommand.php [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+        $provider = $this->providerRepository->findOneByIdentifier($providerIdentifier);
+        if ($provider === null) {
+            $io->error(sprintf('No provider record has the identifier "%s".', $providerIdentifier));
+
+            return Command::FAILURE;
+        }
+
+        $stream = $this->resolveInputStream($input);
+        if ($stream === null) {
+            $io->error('Refusing to read the key from a terminal. Pipe it in, for example: printf \'%s\' "$KEY" | vendor/bin/typo3 nrllm:provider:set-key ' . $providerIdentifier);
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply Classes/Command/SetProviderApiKeyCommand.php]: Agreed, fixed in cfe5f538. The message now carries a fixed `<provider>` placeholder instead of the argument that was just read. Echoing user input into a line the reader is invited to paste into a shell is a footgun regardless of how narrow the path is, and the identifier added nothing there — the reader typed it a moment ago.

File: Tests/Unit/Command/SetProviderApiKeyCommandTest.php [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+    {
+        $stream = fopen('php://memory', 'r+');
+        self::assertIsResource($stream);
+        fwrite($stream, $stdin);
+        rewind($stream);
+
+        $input = new ArrayInput(['provider' => $provider]);
+        $input->setStream($stream);
+
+        return $command->run($input, $this->output);
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply Tests/Unit/Command/SetProviderApiKeyCommandTest.php]: Fixed in cfe5f538 — the stream is closed in a `finally` around `run()`, so it is released even when an assertion inside the command path throws.

File: Classes/Command/SetProviderApiKeyCommand.php [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+     * @param resource $stream
+     */
+    private function readSecret($stream): string
+    {
+        $raw = stream_get_contents($stream);
+
+        // A key piped with `echo` carries a trailing newline; one piped with
+        // `printf '%s'` does not. Everything else is kept verbatim — trimming
+        // further would silently mangle a key rather than reject it.
+        return $raw === false ? '' : rtrim($raw, "\r\n");
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply Classes/Command/SetProviderApiKeyCommand.php:198]: Correct, the comment was wrong about its own code. Fixed in cfe5f538 by making the comment match: all trailing CR/LF are dropped, everything else including surrounding spaces is kept verbatim.

Behaviour stays as it was — that is what nr-vault's own `vault:store --stdin` does, and a key never ends in a line break. The test now pipes `\r\n\n` instead of a single newline, and is renamed to `stripsTrailingLineBreaksButKeepsEveryOtherCharacter` so it states the actual contract.

File: Classes/Command/SetProviderApiKeyCommand.php [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+
+            return Command::FAILURE;
+        }
+
+        $stream = $this->resolveInputStream($input);
+        if ($stream === null) {
+            // The example carries a fixed placeholder rather than the argument
+            // just read: echoing it back into a copy-pasteable shell line would
+            // hand whitespace or metacharacters straight to the next shell.
+            $io->error('Refusing to read the key from a terminal. Pipe it in, for example: printf \'%s\' "$KEY" | vendor/bin/typo3 nrllm:provider:set-key <provider>');
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply Classes/Command/SetProviderApiKeyCommand.php]: Good catch — that placeholder came in with the previous review round and reintroduced a problem one layer down. Fixed in f84cf0fa.

Rather than escaping the brackets, the shell example is gone from the error message. It already exists in `--help`, and an error branch is the wrong place to keep a second copy of usage text that then has to be kept in sync and escaped correctly. The message now says what to do and points at `--help` for the worked line.

That also settles the earlier concern in the same spot: with no shell line in the message, there is nothing to paste and nothing to escape.
